### DISCUSSION TITLE
Export document list as CSV

### DISCUSSION THREAD

**Discussion by mtettke123** [paperless-ngx/paperless-ngx#1796] [Feature Requests]
Story:
I'm actually preparing my tax declaration - the first one since using paperless.
I've all documents sorted and tagged in paperless. So, I can search for all invoices a.e. which works great.

Wish:
As I now have a list of all invoices, an "export as CSV" button would be great, to export the list of all filtered elements. I could just use this list, import it to whatever and would have saved so much time and monkey work!

Wish 2 (much in the future and handled in another request):
If there was a custom field like "amount", it would be even greater....

**Comment by nlspln** [comment]
I would use this to print an inventory sheet to add to each binder so I can look at a glance which documents are in the binder and under what number I can find them (in case I need to do a lookup and I don't have access to the web interface)

**Comment by nodecentral** [comment]
Can something like that be done with the REST API, maybe not a CSV, but a list of files with a defined document type/tag etc ?

**Comment by 23ewrdtf** [comment]
I'm looking for this too. I receive invoices in PDF and I would like to OCR them and then transfer them (tables in them) to csv or excel tables. There is https://github.com/invoice-x/invoice2data but it's clunky.

Maybe we can connect the two?

**Reply by mtettke123** [reply]
While I would really love this feature - I don't think we should put them together in one feature request: What I would love to have here, is just a generation of a CSV from listed items. Nothing fancy, pretty basic.

What you like to achieve - while being a great idea - is muuuch more complex (if I understand it correctly): analything the content of invoices and exporting... I know there are packages for this and DMS that can do this, but this is really complex!

**Comment by 23ewrdtf** [comment]
Yes, I know it's not easy. Been trying to setup invoice2data but the results varies.

For now I can add invoice2data to Post-consumption script.

Detecting tables and basic csv export would be a huge help.

**Reply by mtettke123** [reply]
Just read a little bit into invoice2data. Seems like a fun project and good idea to integrate into paperlesspaperless combined with custom fields (we all are waiting for) to automagically insert a.e. amount of an invoice!

**Comment by 23ewrdtf** [comment]
happy to test. I have lot's of invoices in various types and quality.

**Reply by Frogy76** [reply]
Is it possible to add invoice2data to paperless?

**Comment by pilot479** [comment]
I am another big supporter for this type of feature request please...csv export (ability to filter as well) of labelling records of the documents (titles, types, correspondents, etc).

**Comment by enricohessel** [comment]
I would like to see this feature as well for prepping taxes. 
Best practice:
- Adding the new custom fields to ALL selected documents
- typing in the values if ocr does not work (e.g. price)
- generate a csv, e.g. filename + custom fields + dates

**Reply by georgniemeyer** [reply]
I have the same use case as @enricohessel and would really appreciate an easy way to generate an overview and perhaps even a sum of the custom field "invoice amount" for selected documents. That would make the annual tax declaration much easier.

**Reply by Radnabe** [reply]
Me too

**Comment by mtettke123** [comment]
If it only was possible to fund something like this!

**Reply by dast5022** [reply]
I also would like to see that function in the ui of paperless-ngx. In the meantime, perhabs a little python script could help: 
```
from ast import Slice, Try
import csv
from inspect import cleandoc
from xml.etree.ElementTree import tostring
import requests
import yaml
import sys


def load_config():

    # Change your values here:
    config={}
    config['token']="yourtoken"
    config['api_url']="http://your-ip/api/"
    config['search_string']="tags__id__all=10"
    config['result_size']=1000
    config['delimiter_fields']=";"
    config['delimiter_tags']= ","
    config['output_file']="output.csv"

    return config


def load_enums(api_url, token, keyfield):
    headers = {'Authorization': f'Token {token}'}
    response = requests.get(api_url+"?page_size=10000", headers=headers)
    #print(response)
    enums = response.json()
    filtered_enums = {}
    for k in enums["results"]:
        filtered_enums[k["id"]] = k[keyfield]

    return filtered_enums


def load_documents(api_url, token, search_string, size):
    headers = {'Authorization': f'Token {token}'}
    url = api_url+"?"+str(search_string)+"&page_size="+str(size)
    print("trying", url)
    response = requests.get(url, headers=headers)
    document = response.json()

    if "results" in document and document["count"] >= 1:
        return document["results"]
    else:
        raise ExNoDocument(search_string)


def replace_field_value(value, mapping, categorie):
    if value != "" and value != None:
        try:  
            return mapping[value]
        except:
            print(f'Fehler: "{value}" in {categorie} nicht gefunden.')
            raise ExNoItem(str(value))
    else:
        return ""

class ExNoDocument(Exception):
    def __init__(self, filename):
        self.message = "No document found in paperless"
        self.filename = filename
        super().__init__(self.message, self.filename)    

class ExNoItem(Exception):
    def __init__(self, itemname):
        self.message = "item not found in paperless"
        self.itemname = itemname
        super().__init__(self.message, self.itemname)    

def main():
 
    config=load_config()
    token = config['token']

    # read enums from paperless
    try:
        correspondents = load_enums(config["api_url"]+"correspondents/", token, "name")
        document_types = load_enums(config["api_url"]+"document_types/", token, "name")
        paths= load_enums(config["api_url"]+"storage_paths/", token, "name")
        tag_list= load_enums(config["api_url"]+"tags/", token, "name")
    except ExNoItem:
        print("replacing enum failed")
        exit()

    # read documents from paperless
    try:
        documents = load_documents(config["api_url"]+"documents/", token, config["search_string"], config['result_size'])
    except ExNoDocument as e:
        print("No matching document was found")
        exit()

    # loop through all documents
    documents_out=[]
    i=0
    for doc in documents:

        # replace enums
        doc['correspondent'] = replace_field_value(doc['correspondent'], correspondents, "correspondents")
        doc['document_type'] = replace_field_value(doc['document_type'], document_types, "document_types")
        doc['storage_path'] = replace_field_value(doc['storage_path'], paths, "paths")
        
        # replace tags
        tags_in=[]
        tags_out=[]
        tags_in=doc['tags']
        for tag in tags_in:
            tag = replace_field_value(tag, tag_list, "tags")
            tags_out.append(tag)
        
        # transfer fields to destination list
        row={}
        row['id'] = doc['id']
        row['original_file_name'] = doc['original_file_name']
        row['correspondent'] = doc['correspondent']
        row['storage_path'] = doc['storage_path']
        row['document_type'] = doc['document_type']
        row['created_date'] = doc['created_date']    
        row['title'] = doc['title']
        row['tag_list']=config['delimiter_tags'].join(tags_out)
        documents_out.append(row)
        i+=1
    # export to csv
    csv_file_path = config['output_file']
    print("found "+str(i)+" lines and trying to write to "+csv_file_path)
    
    fieldnames = ['id','original_file_name', 'storage_path', 'correspondent', 'document_type', 'title', 'tag_list', 'created_date']
    
    with open(csv_file_path, 'w', newline='') as csv_file:
        writer = csv.DictWriter(csv_file, fieldnames=fieldnames, delimiter = config['delimiter_fields'])
        writer.writeheader()
        writer.writerows(documents_out)

if __name__ == '__main__':
    main()
```

**Reply by mtettke123** [reply]
just tested (path mounted into paperless-docker), works. Thank you! :-)

Do you have an idea how to add custom fields to the export?

**Reply by mtettke123** [reply]
Answering to myself, as solved with the help of ChatGPT and the [API documentation](https://docs.paperless-ngx.com/api/):

adding field ID of dynamic field to config (guessed the ID...)
`    config['custom_field_id_for_betrag']=1`

in main, when looping through the documents:
```
    for doc in documents:

        # custom fields
        betrag_value = None

        # Check if 'custom_fields' exists in the document
        if 'custom_fields' in doc:
            # Iterate through each custom field to find 'Betrag'
            for custom_field in doc['custom_fields']:
                if custom_field['field'] == config['custom_field_id_for_betrag']:
                    betrag_value = custom_field['value']
                    break  # Stop searching once 'Betrag' is found
```

finally adding the row
`        row['Betrag']=betrag_value if betrag_value is not None else ''`

and the index:
`    fieldnames = ['id','original_file_name', 'storage_path', 'correspondent', 'document_type', 'title', 'tag_list', 'created_date', 'Betrag']`


"works for me".

**Reply by docb7** [reply]
Thank you very much, that was exactly what I was looking for. I also would like to add media_filename to the output, but I just get error messages when trying to add that (to the enums block or in the row block) - as you might already have guessed, I have no clue about api but I guess, as I tried to understand the documentation, the media_filename value is under a different path (/api/documents/<id>/metadata/) while all other values are under /api/documents/<id>/. But I do not know, how to change that. It would be great if @dast5022 or you could point me in the right direction.

**Reply by simonseres** [reply]
Thanks @dast5022  I was using your script as well! 
Works out of the box once I changed "tags__id__all=10" to "tags__id__all=1".
Tbh, I googled around and tried to understand this search query in the Rest API description of paperless as well as on Django docs, but I didn't quite get it.
Can you lead me to understand what the search query actually does?

Also thanks to @mtettke123 - i was inspired by what you did and enhanced the code as well for my individual needs! :)

**Reply by dast5022** [reply]
Hi, @simonseres, the query "tags__id__all=x" selects all documents which are tagged with the specific tag x. For x you have to fill in the id of that tag, not the name. You can find that id via the Django UI. When opening the page for a single tag there, you can find the id in the URL of that page. Don´t ask me what this suffix "_all" has to do with, it also works without.

**Reply by shamoon** [reply]
If you play with the UI or look at the filters in the api browser you will see `__all` means has all the tags (“AND”) so `__in` is “OR”. Of course, both only are relevant with more than one tag.

**Comment by andi0b** [comment]
I'm also missing this feature.

### Some additional thoughts:

1. I would like to include this document list into the existing multi document ZIP file export functionality. 
The document list should be part of the ZIP file, one CSV in the root directory should be enough. There should be a column with a relative file path to the document.

2. It should support not only CSV, but also JSON and JSONL (json lines), and preferrably all formats can be added to the export at the same time

3. CSV separator should be locale (language) specific or changable  
In some locales the separator is a semicolon, because the decimal seperator is a comma. This is also called "European format" and an often missed crucial requirement from US based developers. European Excel versions for example can't simply open comma separated CSV files, this often causes big issues if you send it to non-tech people (simple double-clicking doesn't work). There is an example on the [Wikipedia page about CSV](https://en.wikipedia.org/wiki/Comma-separated_values#Example).

4. Custom fields should be included, maybe notes too.

**Comment by bembembene** [comment]
mtettke123 how did u mount the path in docker? greetings ben

**Comment by schubidoo** [comment]
I am also very strongly wishing for that feature. I prepared all my stuff assuming the possibility exists, which was a bit naive, my bad :-) I tried to work around it by selecting the text in the browser and converting the text to a table, but that does not work, the table text is not selectable - anyone know why, and how to change that?

Working around this by OCRing screenshots just feels soooo stupid :-)

**Comment by nodecentral** [comment]
Is a dB export an option for anyone to explore..

https://stackoverflow.com/questions/1120109/how-to-export-table-as-csv-with-headings-on-postgresql

**Comment by SpuriousGer** [comment]
Doing it by connecting to the PostgresSQL database is very simple, but probably prohibitive to many.

Using the API would be an easy way in my opinion. You could get the JSON and transform it to CSV (or another format).

I think a native export with certain columns would be cool, but I think it would get a bit messy with custom fields potentially.

**Comment by dast5022** [comment]
I´ve build a template for MS Excel with VBA to download all informations of documents via API. This could be hopefully also used by persons without further programming skills. I´d be happy about feedback. See here: https://github.com/dast5022/paperless-dde

**Reply by mtettke123** [reply]
> I´ve build a template for MS Excel with VBA to download all informations of documents via API. This could be hopefully also used by persons without further programming skills. I´d be happy about feedback. See here: https://github.com/dast5022/paperless-ngx-document-importer

Works for me, is a little bit easier to handle than the python script, as soon as you are keen enough to enable macros.
Thank you for your work!
Custom tags don't work yet, as it seems?

Have a good rest-of-the-year!

**Reply by dast5022** [reply]
Hi [mtettke123](https://github.com/mtettke123), now it should also work with custom fields. Give it a try, see release v1.2.0

**Comment by mjrcrm** [comment]
I receive the following error after entering the auth token:

![Screenshot 2025-04-20 210938](https://github.com/user-attachments/assets/e4145a25-81db-48c4-95ea-e4b752d97b91)

I understand that this may be a network problem, but do not understand the mechanics as all paperless-ngx functions seem to work flawlessly - using version 2.15-2.


Debugger points to this:

 ' Make the first API request
    Debug.Print queryUrl
    Set http = CreateObject("MSXML2.XMLHTTP")
    **http.Open "GET", queryUrl, False**
    http.setRequestHeader "Authorization", "Token " & UserToken
    http.send

Any help would be much appreciated.

**Comment by McTristan** [comment]
I’m currently working on a small Web API project based on .NET 8. Its purpose is to export documents and their metadata as a ZIP file, including an Excel spreadsheet with nicely formatted metadata.

The API already supports filtering by tags (both include and exclude), document type, and custom fields. A second endpoint allows you to specify a saved view by its ID; the API then applies the configured filters and sort order from that view and generates the Excel file accordingly.

Although it’s still in an early stage, it’s already working quite well. I’m planning to add more documentation, set up a build pipeline, and implement an MCP server that can interact with Claude (or Ollama) to generate the export ZIP in a more “natural” way.

The idea is to run this export API as a Docker container alongside a Paperless instance (would work on ARM, AMD64 etc.). You can then trigger the various API functions using curl, via Swagger, or potentially through a small UI I’m considering building.

Does this sound like something you’d be interested in using? If there’s enough interest, I’d be happy to invest more time and release it as open source.

**Comment by McTristan** [comment]
You can find the repository here:

https://github.com/McTristan/ExportPaperless

Please note:

- there is not much documentation available
- developing it took a lot of time and effort so behave nicely ;)
- I developed it because I needed the functionality
- there is no Frontend yet so you need to be familiar with call APIs with curl, wget, Postman or whatever
- I'm planning to integrate with HomeAssistant and/or NodeRed and create a small MCP Server to be used with Ollama and Claude AI.

The rest does not need to be discussed here as my project is ONE solution but not THE solution - use the discussion tab or issue board inside the repository.

You can buy me a coffee if you find the project helpful! Thanks!

**Comment by alpha754293** [comment]
I would LOVE it if there is a way to export the custom fields to a CSV or something.

**Reply by simonseres** [reply]
This Python Script does the job. There are 2 custom fields in it as an example. Credits to @dast5022 and @mtettke123. I just combined it for my use case.
```
from ast import Slice, Try
import csv
from inspect import cleandoc
from xml.etree.ElementTree import tostring
import requests
import yaml
import sys


def load_config():

    # Change your values here:
    config={}
    config['token']="your_token"
    config['api_url']="https://yourinstance/api/"
    config['search_string']="tags__id__all=1"
    config['result_size']=1000
    config['delimiter_fields']=";"
    config['delimiter_tags']= ","
    config['output_file']="output.csv"
    # custom field id's
    config['custom_field_id_for_betrag']=3
    config['custom_field_id_for_lieferant']=5


    return config


def load_enums(api_url, token, keyfield):
    headers = {'Authorization': f'Token {token}'}
    response = requests.get(api_url+"?page_size=10000", headers=headers)
    #print(response)
    enums = response.json()
    filtered_enums = {}
    for k in enums["results"]:
        filtered_enums[k["id"]] = k[keyfield]

    return filtered_enums


def load_documents(api_url, token, search_string, size):
    headers = {'Authorization': f'Token {token}'}
    url = api_url+"?"+str(search_string)+"&page_size="+str(size)
    print("trying", url)
    response = requests.get(url, headers=headers)
    document = response.json()

    if "results" in document and document["count"] >= 1:
        return document["results"]
    else:
        raise ExNoDocument(search_string)


def replace_field_value(value, mapping, categorie):
    if value != "" and value != None:
        try:  
            return mapping[value]
        except:
            print(f'Fehler: "{value}" in {categorie} nicht gefunden.')
            raise ExNoItem(str(value))
    else:
        return ""

class ExNoDocument(Exception):
    def __init__(self, filename):
        self.message = "No document found in paperless"
        self.filename = filename
        super().__init__(self.message, self.filename)    

class ExNoItem(Exception):
    def __init__(self, itemname):
        self.message = "item not found in paperless"
        self.itemname = itemname
        super().__init__(self.message, self.itemname)    

def main():
 
    config = load_config()
    token = config['token']

    # read enums from paperless
    try:
        correspondents = load_enums(config["api_url"]+"correspondents/", token, "name")
        document_types = load_enums(config["api_url"]+"document_types/", token, "name")
        paths= load_enums(config["api_url"]+"storage_paths/", token, "name")
        tag_list= load_enums(config["api_url"]+"tags/", token, "name")
    except ExNoItem:
        print("replacing enum failed")
        exit()

    # read documents from paperless
    try:
        documents = load_documents(config["api_url"]+"documents/", token, config["search_string"], config['result_size'])
    except ExNoDocument as e:
        print("No matching document was found")
        exit()

    # loop through all documents
    documents_out=[]
    i=0
    for doc in documents:

        # replace enums
        doc['correspondent'] = replace_field_value(doc['correspondent'], correspondents, "correspondents")
        doc['document_type'] = replace_field_value(doc['document_type'], document_types, "document_types")
        doc['storage_path'] = replace_field_value(doc['storage_path'], paths, "paths")
        
        # replace tags
        tags_in=[]
        tags_out=[]
        tags_in=doc['tags']
        for tag in tags_in:
            tag = replace_field_value(tag, tag_list, "tags")
            tags_out.append(tag)

        
        # Check if 'custom_fields' exists in the document
        betrag_value = None

        if 'custom_fields' in doc:
            # Iterate through each custom field to find 'Betrag'
            for custom_field in doc['custom_fields']:
                if custom_field['field'] == config['custom_field_id_for_betrag']:
                    #print("DOC", doc.get("id"), "custom_field", custom_field.get("field"), "value=", custom_field.get("value"))
                    val = custom_field.get('value')
                    if val: 
                        betrag_value = float(val[3:]) #string slicing: removing 'EUR' by removing first three letters
                        betrag_value = f"{betrag_value:.2f}".replace('.', ',')
                    else:
                        betrag_value = ""
                    break  # Stop searching once 'Betrag' is found

        # Check if 'custom_fields' exists in the document
        lieferant_value = None

        if 'custom_fields' in doc:
            # Iterate through each custom field to find 'Betrag'
            for custom_field in doc['custom_fields']:
                if custom_field['field'] == config['custom_field_id_for_lieferant']:
                    lieferant_value = custom_field['value']
                    break  # Stop searching once 'Lieferant' is found

        
        # transfer fields to destination list
        row={}
        row['id'] = doc['id']
        row['original_file_name'] = doc['original_file_name']
        row['correspondent'] = doc['correspondent']
        row['storage_path'] = doc['storage_path']
        row['document_type'] = doc['document_type']
        row['created_date'] = doc['created_date']    
        row['title'] = doc['title']
        row['tag_list']=config['delimiter_tags'].join(tags_out)
		# custom field 'rows'
        row['Betrag']=betrag_value if betrag_value is not None else ''
        row['Lieferant']=lieferant_value if lieferant_value is not None else ''

        documents_out.append(row)
        i+=1
        
    
    # export to csv
    csv_file_path = config['output_file']
    print("found "+str(i)+" lines and trying to write to "+csv_file_path)
    
    fieldnames = ['id','original_file_name', 'storage_path', 'correspondent', 'document_type', 'title', 'tag_list', 'created_date', 'Betrag', 'Lieferant']
    
    with open(csv_file_path, 'w', newline='') as csv_file:
        writer = csv.DictWriter(csv_file, fieldnames=fieldnames, delimiter = config['delimiter_fields'])
        writer.writeheader()
        writer.writerows(documents_out)

if __name__ == '__main__':
    main()

```

**Reply by alpha754293** [reply]
Oh awesome! Thank you!!!

(Now I am waiting for `paperless-ai` to fix something with how it is parsing the JSON so that it would actually be able to fill in those fields. They said that the fix should have been implemented, but I am waiting for the docker image to update.)

Thanks.

**Comment by drakarah** [comment]
Wouldn't a simple "Export current view to csv" work? Then the column setup is already done in the UI and WYSIWYG.

Right now I can't even select the table to copy paste it to excel, i have to modify the css in the developer tools of my browser so I can select and copy, feels like jumping through a lot of hoops even though the data I want is right in front of me

**Comment by steffenglock** [comment]
> Story: I'm actually preparing my tax declaration - the first one since using paperless. I've all documents sorted and tagged in paperless. So, I can search for all invoices a.e. which works great.
> 
> Wish: As I now have a list of all invoices, an "export as CSV" button would be great, to export the list of all filtered elements. I could just use this list, import it to whatever and would have saved so much time and monkey work!
> 
> Wish 2 (much in the future and handled in another request): If there was a custom field like "amount", it would be even greater....

Have a look at this:
https://github.com/steffenglock/Paperless-ngx-Export-
https://hub.docker.com/r/steffenglock/paperless-ngx-export

Maybe this can be a quick solution

**Comment by Spaghettio73** [comment]
Still open after four years, so I built it. paperless-export gives you whatever a filter or a saved view shows on screen, as CSV, XLSX, JSON or JSONL. Custom fields come out as columns, you can ask for a totals row, and `--pack` drops the matching PDFs in a folder next to the sheet.

The part I actually cared about: it asks the API how many documents match and refuses to finish unless the row count agrees. No more sheets that are quietly too big because a filter rule got dropped somewhere.

Read-only, token stays in your environment, Python 3.10+, no Docker, MIT. https://github.com/Ildana-ai/paperless-export
